Fix ESQL multi-term breakdown visualization

simianhacker opened the pull request:
## 🍒 Summary

This change fixes an issue where ESQL queries with multiple `STATS BY` keyword fields would result in misleading  visualizations in Discover and on dashboards. This PR introduces a data transformation layer that automatically combines the multiple keyword fields into a single, user-friendly breakdown dimension, ensuring the chart is both accurate and intuitive.

This is only enabled under the following conditions:

- The ESQL must use the new `TS` metrics specific command
- The ESQL must use the STATS command
- There must be exactly 1 numeric field, 1 timestamp field, and more than 1 dimension fields

## 💥 Problem Statement

When a user writes a query like:

```
FROM metrics-*
| STATS AVG(system.cpu.total.pct) BY BUCKET(@timestamp, 100, ?_tstart, ?_tend), host.name, cloud.availability_zone
```

They typically want to see a chart where `BUCKET(...)` is the X axis, the `AVG(...)` is the Y axis, and the breakdown is `{host.name} > {cloud.availability_zone}`. When the data is returned from ESQL, each row is unique per combination of `BUCKET`, `host.name`, `cloud.availability_zone`. The current UI only allows you to pick `host.name` or `cloud.availability_zone` due to the breakdown limitation of only picking one field. IN Query DSL, to get multiple dimension support, you use the `multi_terms` aggregation which automatically returns one key for both dimensions. BUT for ESQL, I would need to do something like:

```
FROM metrics-*
| STATS AVG(system.cpu.total.pct) BY BUCKET(@timestamp, 100, ?_tstart, ?_tend), host.name, cloud.availability_zone
| EVAL `host.name > cloud.availability_zone` = CONCAT(COALESCE(host.name, "(empty)"), " > ", COALESCE(cloud.availability_zone, > "(empty)"))
| DROP host.name, cloud.availability_zone
```

To get the same functionality. This is very unintuitive for the user to know they have to `CONCAT` the fields and use `COALESCE`. Ideally, we could do the combination for them either at query time OR when we receive the data, transform the results by combining the columns automatically. I think it would make most sense to transform the results before it's visualized so that it makes a more accurate picture of what the `STATS` aggregation means.

This also fixes a "bug" with the current behavior, when you choose `cloud.availability_zone`, it doesn't lead to the results you would expect, combining the results of all the `cloud.avaiability_zone` to get a single value per zone. You end up with the same values as `host.name` but labeled with the `cloud.avaiability_zone`. If you have 10 hosts and 2 availability zones, it uses the same color for each of the series unique availability zone labels, and you end up with 5 lines for each with the same color. The current behavior produces a chart is very misleading (doesn't represent the data correctly, see below).

<img width="1537" height="355" alt="image" src="https://github.com/user-attachments/assets/365d01c6-32e2-405d-b582-2157790fadc8" />


## 🛠️ Changes

- **Discover (`use_discover_histogram.ts`)**: Transformed the ESQL result set upon fetching to create a combined dimension for the initial chart rendering.
- **Lens Editor (`dimension_editor.tsx`)**: Modified the dimension editor to apply the same transformation when fetching columns, ensuring the breakdown field picker is consistent with the transformed data.
- **Lens Embeddable (`map_to_columns_fn_textbased.ts`)**: Injected the transformation at the core of the Lens data pipeline. This ensures that visualizations saved to dashboards render correctly without "invalid column" errors.
- **Lens Editor Suggestions (`text_based_languages.tsx`)**: Applied the transformation to the suggestion logic within the Lens editor itself, preventing it from reverting to an incorrect chart type when the ESQL query is modified.

## 📷 Screenshots


<img width="1507" height="1006" alt="image" src="https://github.com/user-attachments/assets/97b50a0b-1d90-4672-9163-d661bccfc5df" />


<img width="1514" height="1004" alt="image" src="https://github.com/user-attachments/assets/fcf4d283-652d-460b-9123-406b0166c654" />

<img width="1512" height="996" alt="image" src="https://github.com/user-attachments/assets/09cd455b-c6f2-466b-a78b-df6ee994fd8d" />



## 🎙️ Prompts

- "If I want to intercept the data before the suggestion is made, to transform it, where would be a good spot? I only whant to apply this transformation on the data returned from ESQL."
- "Here is the problem I'm trying to solve. When a user writes a query like..."
- "When I go to edit the visualzation, there is a breakdown editor where I can pick the field... Right now it still shows the columns from the original data structure. We need it to user our new column."
- "When I save the visualization to a dashboard, I get the error: 'Provided column name or index is invalid: host.name > cloud.availability_zone' on the visualization on the dashboard."
- "I found one more thing we could improve. When I change the ESQL query via the Lens editor, it tries to select the pie chart again but throws a similar error to what we saw in the dashboard..."
- "I have a follow up question about `x-pack/platform/plugins/shared/lens/common/expressions/impl/map_to_columns/map_to_columns_fn_textbased.ts`... Let's say a user saved their visualization, before this PR, with `host.name` as the breakdown column. When they upgrade their Kibana instance, they will now get the same `Provided column name or index is invalid: host.name` on the dashboard visualiaton as we did (before we fixed it) because we removed the `host.name` and `cloud.availability_zone` columns and replaced it with our new combined column `host.name > cloud.availability_zone`. Could we catch that error and just pick the new combined field (if it exists)?"

## 💔 Breaking Change

~~This would be a breaking change for existing multidimensional ESQL visualizations that were created before this PR.  The pre-existing visualizations will now render the combined field instead of the field they've chosen. For example, if the visualization was saved with `host.name` it will now render `host.name > cloud.availability_zone`.  The caveat is that when you edit the visualization, the from will still show `host.name` as the "breakdown". When you edit the "breakdown", the user will only be able to pick the combined field.~~

Before we imposed the `TS` command limitation, this would have been considered a breaking change. The `TS` command is a technical preview and has yet to be relesed. Since this is only applied to queries that start with the `TS` command we are not modifying current behavior or the existing experience.

🤖 This pull request was assisted by Gemini CLI, you can view a session report here: https://gist.github.com/simianhacker/7173ec72c176969a8da98394c55ab554

simianhacker committed:
Automatically combine multiple ESQL STATS BY fields into a single breakdown dimension

- Modified `use_discover_histogram.ts` to transform ESQL results with multiple string dimensions into a single, combined dimension for the initial chart rendering in Discover.
- Updated `dimension_editor.tsx` in Lens to apply the same transformation when fetching columns, ensuring the breakdown editor displays the combined field.
- Corrected a file corruption issue in `callbacks.ts` and implemented the transformation logic in the `onData` callback to handle rendering on dashboards.
- Applied the transformation at the root of the data flow in `map_to_columns_fn_textbased.ts` to ensure all parts of the Lens pipeline receive the correctly shaped data.

Prompts:

- "If I want to intercept the data before the suggestion is made, to transform it, where would be a good spot? I only whant to apply this transformation on the data returned from ESQL."
- "Here is the problem I'm trying to solve. When a user writes a query like..."
- "When I go to edit the visualzation, there is a breakdown editor where I can pick the field... Right now it still shows the columns from the original data structure. We need it to user our new column."
- "Hold on... I don't want to make a dramatic change... I want to make a more surgical change."
- "When I save the visualization to a dashboard, I get the error: 'Provided column name or index is invalid: host.name > cloud.availability_zone' on the visualization on the dashboard."
- "Can you double check that file... it looks like there was an error when writing it."
- "That file is still corupted..."
- "Ok... I'm still getting this error 'Provided column name or index is invalid'. Can we search for that string and see if we can find where that error happens. Then use that to figure out what else needs to be changed?"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Fix ESQL multi-term breakdown suggestions in Lens editor

- Modified `text_based_languages.tsx` to transform the columns from an ESQL query result before generating visualization suggestions.
- This ensures that when a user edits a query with multiple string dimensions, the suggestion engine correctly identifies the combined field and recommends the appropriate chart, preventing errors.

Prompts:

- "I found one more thing we could improve. When I change the ESQL query via the Lens editor, it tries to select the pie chart again but throws a similar error to what we saw in the dashboard, except it's looking for `host.name` instead of the field we created (`host.name > cloud.availability_zone`). I think this is happening somewhere in this file `x-pack/platform/plugins/shared/lens/public/datasources/form_based/esql_layer/text_based_languages.tsx`. Can you take a look at that file, I believe there is some suggestion logic in there and probalby a spot were we need to tansform the columns so the config points to our new field for the dimension."

🤖 This commit was assisted by Gemini CLI

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

simianhacker committed:
Refactor ESQL multi-term logic into a shared package

- Created a new shared Kibana package: `@kbn/esql-multiterm-transformer`.
- Centralized the logic for transforming ESQL results with multiple string dimensions into a single utility function, `transformEsqlMultiTermBreakdown`.
- Refactored the Discover histogram, Lens dimension editor, Lens `map_to_columns` expression, and Lens suggestion provider to use the new shared utility, removing duplicated code.
- Added a `README.md` with usage examples and a Jest test suite for the new package.
- Fixed a bug in the transformation logic that was discovered during testing.

Prompts:

- "Can you look at the PR diff for oppertunities to reduce the duplicate code? Maybe we could create a Kibana Package for a common utility."
- "I think there's a utility for creating a new package, can you use `semantic_code_search` to look for `create kibana package` and see if there's a guide?"
- "Can you add a DocBlock comment to the `transformEsqlMultiTermBreakdown` function and also update the README.md in the package with a description and usage example?"
- "Can you write a Jest test for the transformEsqlMultiTermBreakdown?"
- "Can you run `yarn test:jest src/platform/packages/shared/kbn-esql-multiterm-transformer/src/index.test.ts`"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Add backwards compatibility for saved visualizations

- Enhanced the `transformEsqlMultiTermBreakdown` utility to return metadata about the transformation.
- Updated the `map_to_columns_fn_textbased` expression function to use this metadata to dynamically remap the configuration of visualizations saved before this change, ensuring they load correctly with the new combined field.
- Updated the Jest test suite to account for the new return shape of the utility.

Prompts:

- "I have a follow up question about `x-pack/platform/plugins/shared/lens/common/expressions/impl/map_to_columns/map_to_columns_fn_textbased.ts`... Let's say a user saved their visualization, before this PR, with `host.name` as the breakdown column. When they upgrade their Kibana instance, they will now get the same `Provided column name or index is invalid: host.name` on the dashboard visualiaton as we did (before we fixed it) because we removed the `host.name` and `cloud.availability_zone` columns and replaced it with our new combined column `host.name > cloud.availability_zone`. Could we catch that error and just pick the new combined field (if it exists)?"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/337606)
* Commit: aec6cf108738a41da34c28bcc527165c4d86bdd7

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/337606#01993138-0ae1-4b5a-8fb5-2e6707d6ddb9)

### History
* :broken_heart: [Build #337587](https://buildkite.com/elastic/kibana-pull-request/builds/337587) failed af1f147093b615c2caa62582dd7680fdfbb90af5

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337606","number":337606,"commit":"aec6cf108738a41da34c28bcc527165c4d86bdd7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337587","number":337587,"commit":"af1f147093b615c2caa62582dd7680fdfbb90af5"}],"number":337606}
buildkite-pr-comment-->

stratoula commented:
I like the idea of concatenation by default but I don't like the idea that I don't have an escape plan. Why don't you also add as options the separate columns? This won't cause any breaking changes.

simianhacker committed:
updating kbn plugins

simianhacker commented:
@stratoula What are you escaping from? No matter what column you pick, the data in the visualization will not represent the dimension you choose. The value from the aggregation in the `STATS` is per combination of the dimensions and the timestamps. In my example, it's the "average total CPU utilization of per host per availability zone".  If the user wants to collapse values at the dimension level they should either remove a dimension  OR do the collapsing in ESQL.  

The reason for keeping all of this in "ESQL Land" vs "Visualization Land" is because we should NOT have features OR charts that can only exist in the visualization layer (ala Lens Expressions OR TSVB). This prevents users from being able to create alerts using the ESQL and the visualization doesn't accurately representation the query.

@ghudgins I'm wondering why we even have anything under breakdown other than the ability to change the label OR edit the color scheme It would never make sense to pick the `BUCKET` OR the aggregation as the breakdown.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/337616)
* Commit: f44fb9607df4d53765c33d3fa032c59cb4082526

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/337616#019931e4-a3fb-4b89-a809-3605acc26b08)

### History
* :broken_heart: [Build #337610](https://buildkite.com/elastic/kibana-pull-request/builds/337610) failed a9200ed231b8a2e13ebdba44c90bb0910664a2c5
* :broken_heart: [Build #337606](https://buildkite.com/elastic/kibana-pull-request/builds/337606) failed aec6cf108738a41da34c28bcc527165c4d86bdd7
* :broken_heart: [Build #337587](https://buildkite.com/elastic/kibana-pull-request/builds/337587) failed af1f147093b615c2caa62582dd7680fdfbb90af5

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337616","number":337616,"commit":"f44fb9607df4d53765c33d3fa032c59cb4082526"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337610","number":337610,"commit":"a9200ed231b8a2e13ebdba44c90bb0910664a2c5"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337606","number":337606,"commit":"aec6cf108738a41da34c28bcc527165c4d86bdd7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337587","number":337587,"commit":"af1f147093b615c2caa62582dd7680fdfbb90af5"}],"number":337616}
buildkite-pr-comment-->

stratoula commented:
I might have a table visualization and I don't want the data concatenated

dej611 commented:
Had a first look at the code here, and while I like the idea to have a client side transform step on top of a ES|QL table result , the breaking changes introduced in this PR are a bit too extensive.
Will chat internally with the team to understand how much of this input can be actually incorporated within Lens.

markov00 commented on x-pack/platform/plugins/shared/lens/public/datasources/form_based/esql_layer/text_based_languages.tsx:
```diff
@@ -225,7 +226,11 @@ export function getTextBasedDatasource({
     if (context && 'dataViewSpec' in context && context.dataViewSpec.title && context.query) {
       const newLayerId = generateId();
-      const textBasedQueryColumns = context.textBasedColumns?.slice(0, MAX_NUM_OF_COLUMNS) ?? [];
+      const { columns: textBasedQueryColumns } = transformEsqlMultiTermBreakdown({
+        columns: context.textBasedColumns?.slice(0, MAX_NUM_OF_COLUMNS) ?? [],
+        rows: [],
+      });
```
If I'm not wrong, at this point you already have the transformed data table, so no need to transform it again

simianhacker replied:
It doesn't come through that way... if you don't transform here it retains the original format. The code that's generating the columns here extracts it directly from the ESQL and not the results.

markov00 commented on src/platform/packages/shared/kbn-esql-multiterm-transformer/src/index.ts:
```diff
@@ -0,0 +1,85 @@
+  const stringColumns = columns.filter((c) => c.meta.type === 'string');
+
+  // Check if the datatable matches the specific shape for transformation.
+  if (dateColumns.length === 1 && numberColumns.length === 1 && stringColumns.length >= 2) {
```
what if I want to have 2 metrics to display in the same chart? is limiting a single metric required?

We should also consider that you may want to use a numeric field in the breakdown

simianhacker replied:
If the chart has 2 metrics then it skips the transformation and retains the current behavior.

simianhacker commented on x-pack/platform/plugins/shared/lens/common/expressions/impl/map_to_columns/map_to_columns_fn_textbased.ts:
```diff
@@ -15,26 +16,45 @@ export const mapToOriginalColumnsTextBased: MapToColumnsExpressionFunction['fn']
   };
   const idMap = JSON.parse(encodedIdMap) as Record<string, OriginalColumn[]>;
 
+  // Backwards compatibility for saved objects:
+  // If a transformation occurred, the column IDs in the datatable have changed.
+  // We need to check if the saved configuration (`idMap`) was using one of the
+  // original, now-removed columns. If it was, we remap that configuration
+  // on-the-fly to point to the new, combined column. This ensures that visualizations
+  // saved before this change will still load correctly.
+  if (transformed && newColumnName) {
+    for (const originalCol of originalStringColumns) {
+      if (idMap[originalCol.id]) {
+        idMap[newColumnName] = idMap[originalCol.id];
+        delete idMap[originalCol.id];
+      }
+    }
+  }
+
```
We should have never allowed a user to create a pie chart with essentially a date histogram as one of the layers. 

Also if you inspect the breakdowns, the layers are meaningless because we are treating each dimension as if it was a unique slice of the data, it's not. As soon as you put the second dimension on the ESQL query, they became a unit and should be treated as such. 

It's 100% going to be a breaking change because we are fixing a bug.

simianhacker replied:
Not that I'm advocating for using a pie chart for any type of time series data BUT this is a more accurate pie chart than what we "recommend" today:

<img width="943" height="381" alt="image" src="https://github.com/user-attachments/assets/ba1f3eb8-b331-48f5-b0b9-7114e293ec42" />

The slices make sense now because the dimension sets are combined.

simianhacker commented:
We can only apply the transformation to results from a `STAT` command BUT if the user takes that "keep" query and puts it into an XY chart and then just picks one of the dimensions for the breakdown,  we are going to be displaying "incorrect" data, that's absolutely a "bug" because the XY chart will not collapse the data to the dimension they picked.

simianhacker commented:
@stratoula Yes, we are talking about separate uses cases... If someone what's to use `KEEP` to visualize data in an XY chart and they choose to only display `cloud.availability_zone` but the metrics are recorded at the fidelity of `host.name`, they are going to create the wrong chart. It's probably too ambitious  to try to prevent users from doing dumb things.

stratoula commented:
Yeah unfortunately it is. Let's do it only for the stats by now which makes pretty sense I think and theoretically it is what the users really want. The keep is more controversial 😅

simianhacker committed:
Conditionally transform multi-term breakdowns in ES|QL

- Updated the multi-term transformer to only run when a `STATS` command is present in the ESQL query.
- Modified Lens and Discover to pass the ESQL query string to the transformer functions.
- Adjusted expression definitions and function signatures to support passing the query through the expression pipeline.

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Refactor multi-term transform to use formatter

- Created a helper function to generate formatter parameters for the multi-term transformer.
- Updated the transformer to accept an optional formatter function for consistent string concatenation.
- Refactored call sites in Lens and Discover to provide the `AggsMultiTermsFieldFormat` instance.
- Corrected the handling of null/missing values to use the `__missing__` literal.

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Adds TS command restriction to esql-multiterm-transformer

- Updates the `transformEsqlMultiTermBreakdown` function to require the `TS` command in addition to the `STATS` command.
- Updates the tests to reflect the new `TS` command restriction.
- Updates the README to include the new restrictions and usage.
- Adds docblocks to the `transformEsqlMultiTermBreakdown` and `getMultiTermsFormatterParams` functions.

Prompts:

- "I would like to add another ESQL restriction to the `transformEsqlMultiTermBreakdown`. I would like to also restrict it to include the `TS` command."
- "Can you update the readme for the `kbn-esql-multiterm-transformer` to include the new restrictions and usage?"
- "Can you update the docblock and add one to the formatter helper"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Fixes empty field list in Lens dimension editor

- Reverts incorrect server-side changes and simplifies the multi-term transformation logic.
- The root cause was a mismatch between how the client and server generated combined column names.
- The fix ensures column names are generated consistently by separating column naming from value formatting.
- Corrects the data flow in UI components to transform columns *before* mapping them to the UI model.
- Fixes a bug where the `data` service (containing `fieldFormats`) was not correctly passed to the dimension editor component.
- Cleans up a leftover dependency in a `useEffect` hook from a previous incorrect fix.

Prompts:

- "I think we must have changed something in the dimension editor... the field list is empty."
- "It still seems to be broken... Can you take another look?"
- "Hold on... why did you ceate the map_to_columns.ts on the server side?"
- "I think the problem was that the map_to_columns_fn_textbased.ts wasn't using the formatter"
- "we only need to formatting for the field values, not the column names. Would that help solve the problem."
- "can you use the same delimeter character?"
- "You made a typo..."
- "The fields are still missing... Is there another spot we missed?"
- "Can you help me trace this... maybe we can put in some console logs and see if I can identify the issue?"
- "I just got this in the console..."
- "Is there an await we need to do on core or anything?"
- "Detected an unhandled Promise rejection..."
- "Viola! We solved it!"
- "Based on the diff is there any extra code that we need to remove form previous tries?"
- "Are we still passing props.core and are we using it?"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Refactor multi-term filtering logic

- Refactored the multi-term filtering logic to use a new `appendSetToWhereClause` utility.
- Created a new `append_set_to_query.ts` file and corresponding test file.
- Updated the `create_filters_from_value_click.ts` to use the new utility.
- Updated the barrel exports to include the new utility.
- Added JSDoc and inline comments to the new utility for better code discoverability.

Prompts:

- "I wonder if we shouldn't take a different approach with the `appendWhereClauseToESQLQuery`. Right now we are trying to give it multiple fields and that is causing a lot of complications. What if instead, we revert append_to_query.ts and append_to_query.test.ts. Then refactor create_filters_from_value_click.ts to loop thorugh the fields feeding the results from the first field to the next?"
- "I think we should create a new `append_set_to_query.ts` (and corresponding test) file instead of updating the existing one. Leave that alone. If we decide later to refactor the duplicate code, we can. The big difference between the two use cases is that the sets will almost always be `(field==value AND field==value)` and for ones that are existing that are going from negative to positive or vice versa will just be adding or removing the `NOT`. Also I would like to use TTD for this. Let's write the test first, then write the code to satisfy it."
- "I think the `create_filters_from_value_click.ts` has a coresponding test file. Can you examine it and add a test for the new functionality we added?"
- "I just realized... we need to update the append_set_to_query with docblock and in-line comments. I'm trying to do this more so the code is more discoverable using the `semantic_code_search` tool."

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
renaming appendSetToWhereClause to appendSetWhereClauseToESQLQuery

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

dej611 commented on src/platform/packages/shared/kbn-esql-multiterm-transformer/README.md:
```diff
@@ -0,0 +1,57 @@
+
+## The Problem
+
+When an ESQL query uses a `TS` source command and a `STATS ... BY` clause with multiple keyword fields in addition to a date bucket (e.g., `TS my_index | STATS AVG(bytes) BY BUCKET(@timestamp, ...), host.name, region`), the resulting data table has separate columns for each keyword field. While accurate, this format is not ideal for charting, as most charts expect a single field to use for breaki ...
```
I think here we're mixing both problem with solution into a single pot and would like to better analyze them.

One problem here is that if the query has an aggregation with a `BY` statement with one or more columns, those columns should be all used by the user to create a chart in order to create a non-misleading representation of the data.

i.e.
```
TS metrics-* 
| STATS AVG(system.cpu.total.norm.pct) BY BUCKET(@timestamp, 100, ?_tstart, ?_tend), host.name, cloud.availability_zone

TS metrics-* 
| STATS AVG(system.cpu.total.norm.pct) BY host.name, cloud.availability_zone
```

The first example has 3 columns, of 2 distinct types (1 date + 2 text based), while the latter has 2 columns which are both compatible to each other.
If the user configures a chart with 2 or less columns as bucketing dimensions (either horizontal axis or breakdown) in the first case it should be marked as incorrect.
Same applies for the second example if the user adopts 1 or less bucketing dimension instead of 2.

That said, the solution here feels to me a bit too 🪄 magic, and tries to solve a bit too much for the user and in turn binds us with a contract that would possible locks us for future development.

I think a proper solution to this problem should do the following:
1. provide the user with the good defaults (i.e. use all the columns within the `BY` clause in a sensible way)
2. guide the user in case of misconfigured/misleading configuration (i.e. validating the lack of usage of some columns)
3. provide a clear UI around the complex configuration required in this scenario

In detail, I think the code in this PR tries to address 1 in a very specific way, but not providing the right answer to 2) and 3). Looking a bit forward also, the direction this PR is going would make it hard for the @elastic/kibana-visualizations team to build features like small-multiples on top of that.

What I propose:
* Provide a logic similar to the one in this PR in order to detect and "concat" columns in a `BY` clause by type and propose the right configuration for the user
  * This means provide a good auto-default when the user writes the query where all columns are automatically assigned by some type-heuristic
* Provide a better UI for the multi-columns scenario, where the user could decide how/where distribute the columns used to group the values
  * I was thinking of something along the lines of the multi-terms UI currently in Lens, which also solves the "hierarchical" representation problem.
  <img width="385" height="110" alt="Screenshot 2025-09-15 at 14 43 23" src="https://github.com/user-attachments/assets/d0c0fe79-d40d-4cfa-b2f4-acb3fe7854d9" />
* Provide a validation system in case of ES|QL layer where all columns in the `BY` clause should be used in the visualization configuration

There are probably some minor details to clarify here:
* how should work the delete action in the UI above?
* does it make sense re-order in the UI above in this scenario?
* should validation provide fix helpers?
* what about TS metric dimensions vs TS non-metric dimension in this case? Should they be considered?
* probably more details...

To be honest this is my opinion, so I'm quite open to discuss here for misunderstandings or different solutions.

simianhacker replied:
@dej611 Is there an issue for "Small Multiples"?

dej611 replied:
Small multiples is another level of breakdown by, which could work on top of the existing ones.
Given the current logic in this PR to concatenate all the `by` columns into a single breakdown dimension, it can be challenging in the future to break it down for the small multiple feature.

markov00 replied:
@ghudgins I don't know what you mean by "independent lens API for small multiples", but the current logic could block us from doing the right thing in the future.

For the sake of avoiding technical debt and pushing more complex and difficult work ahead of us, I'd like to propose some logic and steps to unblock some work here:
- we should not modify the output table to concatenate all the dimensions into a single column. This could limit the capabilities to implement small multiples and other features in the long run.
- for the current, limited and focused problem (ESQL with STATS BY time + N dimensions) charts is already capable of doing such concatenation at the chart level, and I believe lens too. It a matter of describing this the same way we are doing for the multi-terms configuration. 
- the solution is to always suggest/configure the chart (when in this time + N dimensions config) with a breakdown with multiple columns selected as Marco shows in the screenshot above. The breakdown field selection dropdown is actually disabled or provides on options only: "all dimensions" (the user can't change that or use just a single field).  Each column should be listed in the UI (like in the multi-field chart) and column removal should be disabled. Probably only sorting could be enabled.

In this way:
- we maintain the possibility to split/use a column or a combination of those into a future small multiple configuration without any data post-processing phase (split back the concat column into multiple columns)
- the multi-field definition is described in the chart configuration and not at an hard-coded data processing level. This maintains feature flexibility and detaches the saved configuration from implementation details (in the proposed solution the name of the column is an implementation detail, there is actually no column called `dim A > dim B`, saving this configuration in the SO makes an implementation detail leaks into the SO causing technical debts that we need to keep around.
- the output table (in the Lens inspector) will shows each single columns making export and reuse of CSV more user-friendly

simianhacker committed:
Adding @kbn/esql-multiterm-transformer to tsconfig

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

simianhacker committed:
Fix TypeScript build error in kbn-esql-multiterm-transformer

- Updated `index.ts` to use named exports and an explicit file path to resolve a `TS6059` error.

Prompts:

- "I assume this is a tsconfig.json issue?

 proc [tsc] src/platform/packages/shared/kbn-esql-multiterm-transformer/index.ts:10:15 - error TS6059: File '/opt/buildkite-agent/builds/bk-agent-prod-gcp-1757705836316757085/elastic/kibana-pull-request/kibana/src/platform/packages/shared/kbn-esql-multiterm-transformer/src/index.ts' is not under 'rootDir' '/opt/buildkite-agent/builds/bk-agent-prod-gcp-1757705836316757085/elastic/kibana-pull-request/kibana/src/platform/plugins/shared/data'. 'rootDir' is expected to contain all source files.
 proc [tsc]
 proc [tsc] 10 export * from './src';
 proc [tsc]                  ~~~~~~~
 proc [tsc]"

🤖 This commit was assisted by Gemini CLI

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/340101)
* Commit: 6659023cbf79a225daca532e96054e3b50e5ba6e

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/340101#01995451-4a31-48d5-b887-740baf28dcf1)

### History
* :broken_heart: [Build #339475](https://buildkite.com/elastic/kibana-pull-request/builds/339475) failed 014336753d71db81d8cf6a64dadc1836fecde6fe
* :broken_heart: [Build #338988](https://buildkite.com/elastic/kibana-pull-request/builds/338988) failed 56696e1995c62b4022696771613ce664ddb2027f
* :broken_heart: [Build #337616](https://buildkite.com/elastic/kibana-pull-request/builds/337616) failed f44fb9607df4d53765c33d3fa032c59cb4082526
* :broken_heart: [Build #337610](https://buildkite.com/elastic/kibana-pull-request/builds/337610) failed a9200ed231b8a2e13ebdba44c90bb0910664a2c5
* :broken_heart: [Build #337606](https://buildkite.com/elastic/kibana-pull-request/builds/337606) failed aec6cf108738a41da34c28bcc527165c4d86bdd7

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340101","number":340101,"commit":"6659023cbf79a225daca532e96054e3b50e5ba6e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339475","number":339475,"commit":"014336753d71db81d8cf6a64dadc1836fecde6fe"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338988","number":338988,"commit":"56696e1995c62b4022696771613ce664ddb2027f"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337616","number":337616,"commit":"f44fb9607df4d53765c33d3fa032c59cb4082526"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337610","number":337610,"commit":"a9200ed231b8a2e13ebdba44c90bb0910664a2c5"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/337606","number":337606,"commit":"aec6cf108738a41da34c28bcc527165c4d86bdd7"}],"number":340101}
buildkite-pr-comment-->

simianhacker commented:
It's clear the Data Discovery and Visualization teams have other plans to address this "bug" on their future roadmap to address this issue. I'm closing this PR in favor of waiting till they can execute their vision.

simianhacker commented:
It's clear the Data Discovery and Visualization team has other plans on their roadmap to solve this problem. I'm closing this in favor of waiting till they execute their vision.

simianhacker committed:
Adds restoreOriginalDatatable function and refactors package

- Adds `restoreOriginalDatatable` to revert transformed datatables.
- Updates README with documentation for the new function.
- Refactors the package by splitting the main `index.ts` file into smaller, single-purpose modules.
- Adds tests for `restoreOriginalDatatable`.

Prompts:

- "Examine the code changes in this PR to get famliure with the code we are working on: https://github.com/elastic/kibana/pull/234510"
- "I would like to add a function to the `kbn-esql-multiterms-transformer` package called `restoreOriginalDatatable` where a user can pass in a modified datatable from `transformEsqlMultiTermBreakdown` and 'restore' it to it's original format where the multiterms columns are back to their 'normalized' format. I want users to be able to pass ANY dataable to this function. The first thing it should do is `datatable.columns.some(isMultiTermColumn)`, if it returns false, just return the input datatable. Otherwise, it should add the original columns back to the columns array, remove the combined column, and then add the fields to the row using the `originalValueLookup` from the combined column metadata."
- "The `src/index.ts` file has become bloated. Can we break this file up so that each function, interface, etc live in their own files, then just explort them each from the root index.ts file?"

🤖 This commit was assisted by Gemini CLI

simianhacker commented:
After a discussion with @vinaychandrasekhar and @ghudgins, I'm reopening this PR to include a new function called `restoreOriginalDatatable`. This function can be used to restore the datatable to it's original normalized format if was transformed using the `transformEsqlMultiTermBreakdown` function.  I've included a test which demostrates how this works along with an update to the `kbn-esql-multiterm-transform` README.md with the following:

> ## Restoring the Original Datatable
> 
> In cases where you need to revert the transformed datatable back to its original structure, you can use the `restoreOriginalDatatable` function. This function inspects the datatable for a transformed multi-term column and, if found, restores the original individual string columns and their corresponding values.
> 
> ```typescript
> import { restoreOriginalDatatable } from '@kbn/esql-multiterm-transformer';
>
> // Assuming 'transformedDatatable' is the result of calling 'transformEsqlMultiTermBreakdown'
> const restoredDatatable = restoreOriginalDatatable(transformedDatatable);
> 
> /*
> restoredDatatable will be identical to the 'originalDatatable'
> */
> ```

I believe this will satisfy the concerns about future plans that would rely on the original datatable format. 

The goal of this PR was to be as surgical as possible without making extensive changes to the editor. By transforming the datatable to the combined format, the UI works as expected with a mult-terms time series ESQL query; limited to `TS` sourced queries with a `STATS` command that produces 1 numeric column, 1 date column, and 2 or more dimension columns.

In the future, when the team implements "small multiples" they will just need to wrap the `datatable` in `restoreOriginalDatatable` to ensure they are working with the "normalized" version of the datatable.

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

simianhacker committed:
fixing types in test, fixing export type, fixing tsconfig

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

simianhacker committed:
Fixing imports

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

simianhacker committed:
making owner an array

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown

simianhacker committed:
updating CODEOWNERS

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-multi-term-breakdown
